See thread http://lists.busybox.net/pipermail/busybox/2012-March/077459.html
for details. Patch by Denys Vlasenko should probably soon be upstream, so
as of busybox-1.19.5 it should not be necessary anymore.

--- coreutils/printf.c	2011-10-28 18:09:32.000000000 +0200
+++ coreutils/printf.c	2012-03-07 11:55:26.982016907 +0100
@@ -132,13 +132,28 @@ static double my_xstrtod(const char *arg
 	return result;
 }
 
+/* Handles %b */
 static void print_esc_string(const char *str)
 {
 	char c;
 	while ((c = *str) != '\0') {
 		str++;
-		if (c == '\\')
-			c = bb_process_escape_sequence(&str);
+		if (c == '\\') {
+			/* %b also accepts 4-digit octals of the form \0### */
+			if (*str == '0') {
+				if ((unsigned char)(str[1] - '0') < 8) {
+					/* 2nd char is 0..7: skip leading '0' */
+					str++;
+				}
+			}
+			{
+				/* optimization: don't force arg to be on-stack,
+				 * use another variable for that. */
+				const char *z = str;
+				c = bb_process_escape_sequence(&z);
+				str = z;
+			}
+		}
 		putchar(c);
 	}
 }
